src/net/http_network.rs

```diff
@@ -0,0 +1,155 @@
+use crate::{
+    helpers::{
+        fabric::{MessageChunks, Network},
+        Error,
+    },
+    net::{
+        discovery::{peer, PeerDiscovery},
+        MpcHelperClient,
+    },
+    protocol::QueryId,
+};
+use futures::{ready, Sink};
+use pin_project::pin_project;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_util::sync::{PollSendError, PollSender};
+
+/// Wrapper around a [`PollSender`] to modify the error message to match what the [`Network`] trait
+/// requires. The only error that [`PollSender`] will generate is "channel closed", and thus is the
+/// only error message forwarded from this [`Sink`].
+#[pin_project]
+pub struct HttpMessageSink<T> {
+    #[pin]
+    inner: PollSender<T>,
+}
+
+impl<T: Send + 'static> HttpMessageSink<T> {
+    #[must_use]
+    pub fn new(sender: mpsc::Sender<T>) -> Self {
+        Self {
+            inner: PollSender::new(sender),
+        }
+    }
+}
+
+impl<T: Send + 'static> Sink<T> for HttpMessageSink<T>
+where
+    Error: From<PollSendError<T>>,
+{
+    type Error = Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.project().inner.poll_ready(cx)?);
+        Poll::Ready(Ok(()))
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        self.project().inner.start_send(item)?;
+        Ok(())
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.project().inner.poll_flush(cx)?);
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.project().inner.poll_close(cx))?;
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Http implementation of a [`Network`]. Uses channels for both [`Sink`] and [`Stream`]
+/// implementations.
+/// # Panics
+/// if `recv_stream` or `recv_messages` called more than once.
+#[allow(dead_code)] // TODO: WIP
+pub struct HttpNetwork<'a> {
+    peers: &'a [peer::Config; 3],
+    query_id: QueryId,
+    sink_sender: mpsc::Sender<MessageChunks>,
```

Comment:
do you plan to resolve the role of this helper somewhere else? I believe currently helper's role is sent in query string

Reply:
that's true, and it'll be pulled off the query string and used for `ChannelId` inside `MessageChunks`. is that what you mean?

Reply:
Yea receiver will get the role from query string, I wasn't sure where exactly we are going to add it to the request on the sender side

Reply:
the client's `send_messages` fn takes it as arg: https://github.com/private-attribution/ipa/blob/main/src/net/client/mod.rs#L23

if the network is aware what role it is, it can just pass it in there. tho that will have to be added to the `HttpNetwork` struct. i have a PR that implements the send, but I just checked and definitely used the wrong role there. whoops, will fix that: https://github.com/private-attribution/ipa/pull/221/files#diff-08bc7005bd05cd19c26aa4d50474446db8892cd720a69a482a523c2099561bb3R72

Reply:
ok, I fixed that PR to use the correct role